SimG4Core/Application/plugins/OscarMTProducer.cc

```diff
@@ -71,6 +78,24 @@ OscarMTProducer::OscarMTProducer(edm::ParameterSet const& p, const OscarMTMaster
   m_runManagerWorker = std::make_unique<RunManagerMTWorker>(p, consumesCollector());
   m_masterThread = ms;
 
+  // Prepare tokens
+  bool isDD4Hep = p.getParameter<bool>("g4GeometryDD4hepSource");
+  edm::ESGetToken<cms::DDCompactView, IdealGeometryRecord> pDD4Hep;
+  edm::ESGetToken<DDCompactView, IdealGeometryRecord> pDDD;
+  if (isDD4Hep) {
+    pDD4Hep = esConsumes<cms::DDCompactView, IdealGeometryRecord, edm::Transition::BeginRun>();
+  } else {
+    pDDD = esConsumes<DDCompactView, IdealGeometryRecord, edm::Transition::BeginRun>();
+  }
+  edm::ESGetToken<HepPDT::ParticleDataTable, PDTRecord> pPDT =
+      esConsumes<HepPDT::ParticleDataTable, PDTRecord, edm::Transition::BeginRun>();
+  m_masterThread->SetTokens(pDDD, pDD4Hep, pPDT);
```

Comment:
Given that the consumption of the EventSetup products is a property of the `OscarMTMasterThread`, it would be somewhat cleaner to do the `esConsumes()` calls there as well. E.g. here
```cpp
m_masterThread->callConsumes(consumesCollector());`
```
and in `OscarMTMasterThread`
```cpp
void OscarMTMasterThread::callConsumes(edm::ConsumesCollector iC) {
  if (m_pGeoFromDD4hep) {
    m_DD4Hep = iC.esConsumes<edm::Transition::BeginRun>();
  } else {
    m_DDD = iC.esConsumes<edm::Transition::BeginRun>();
  }
  m_PDT = iC.esConsumes<edm::Transition::BeginRun>();
}
```
The `OscarMTMasterThread::m_hasToken` is not really necessary.

Reply:
Done

Reply:
but m_hasToken is left for any case - not dangerous.